### ISSUE TITLE
Custom Modules, using Quill.register() saying cannot read property 'register' of undefined

### ISSUE THREAD

**Issue by giorgio19** [KillerCodeMonkey/ngx-quill-example#6]
Hello again. 
Now that I've gotten the editor functionally working, I wanted to try and create a custom module.

When following the strict QuillJS documentation failed, I started working from your ngx-quill example again. While I got close I ran into problems when actually trying to do anything with importing, registering or defining the module. I couldn't get ImageResize to work on my end even though I did install the module in node_modules. The import simply would not recognize the ImageResize, so the rest of functionality could not be implemented or even really attempted.

I then turned to your counter example as it was more closely related to what I wanted to actually do. However, even with your example which shows exactly what to type and the QuillJS documentation I cannot get it to work for the life of me. I got it to work so that the import line worked and everything seemingly compiles in command prompt. But the page itself doesn't actually display anything when I go in my browser, so obviously something is incorrect. I've looked at other people opening issues and tried working through their supposed solutions but just can't get it. After a couple days, I thought I would just ask because it seems like you know how to simply do it and I am most likely missing something pretty dumb. This is everything I have that works and displays a working page in the browser. It may have something to do with parchment perhaps I also could not get that to work for me.

So sorry for the pestering, any help would greatly appreciated.  Just want to say thank you and that this repository I have to say is pretty darn amazing sir. 

![capture](https://user-images.githubusercontent.com/35510706/36628060-f8878ee8-1900-11e8-890b-f0cc69825bf8.PNG)

editor.component.ts
```
import {Component, ElementRef, OnInit, ViewChild, ViewEncapsulation, ViewContainerRef, ComponentFactoryResolver} from '@angular/core';
import {FormBuilder, FormControl, FormGroup, Validators} from '@angular/forms';
import {Observable} from 'rxjs/Observable';

import {QuillEditorComponent} from 'ngx-quill/src/quill-editor.component';
import {AutocompleteBoxComponent} from '../autocomplete-box/autocomplete-box.component';

import 'rxjs/add/operator/debounceTime';
import 'rxjs/add/operator/distinctUntilChanged';

import Quill from 'quill';

// import Counter from './counter';
// Quill.register('modules/counter', Counter);

// // // add image resize module
// import ImageResize from 'quill-image-resize-module';
// Quill.register('modules/imageResize', ImageResize);

// // override p with div tag
// const Parchment = Quill.import('parchment');
// const Block = Parchment.query('block');


@Component({
  selector: 'app-editor',
  templateUrl: './editor.component.html',
  styleUrls: ['./editor.component.scss'],
  encapsulation: ViewEncapsulation.None
})


export class EditorComponent implements OnInit {

  title = '<ul><li>I am example content</li><li><u>And this, too</u></li></ul>';
  isReadOnly = false;
  placeholder = 'placeholder';
  form: FormGroup;
  modules = {};

  constructor(fb: FormBuilder) {
    this.form = fb.group({
      editor: ['test']
    });

    this.modules = {
      formula: true,
      toolbar: [['formula'], ['image']]
    };
  }

  @ViewChild('editor') editor: QuillEditorComponent;

  ngOnInit() {
    this.form
      .controls
      .editor
      .valueChanges
      .debounceTime(400)
      .distinctUntilChanged()
      .subscribe(data => {
        console.log('native fromControl value changes with debounce', data);
      });

    this.editor
      .onContentChanged.debounceTime(400)
      .distinctUntilChanged()
      .subscribe(data => {
        console.log('view child + directly subscription', data);
      });
  }


  setControl() {
    this.form.setControl('editor', new FormControl('test - new Control'));
  }

  setFocus($event) {
    $event.focus();
  }

  patchValue() {
    this.form.controls['editor'].patchValue(`${this.form.controls['editor'].value} patched!`);
  }

  toggleReadOnly() {
    this.isReadOnly = !this.isReadOnly;
  }

  logChange($event: any) {
    console.log($event);
  }

  logSelection($event: any) {
    console.log($event);
  }

}
```
editor.component.html
```
<h3>Default editor</h3>
<quill-editor [style]="{height: '200px'}" (onEditorCreated)="setFocus($event)"></quill-editor>

<h3>Reactive Forms and patch value</h3>
<button type="button" (click)="hide=!(!!hide)">hide / show</button>
<form [formGroup]="form">
  {{form.controls.editor.value}}
  <button type="button" (click)="patchValue()">patchValue</button>
  <button type="button" (click)="setControl()">setControl</button>

  <quill-editor #editor [style.display]="hide ? 'none' : 'block'"
                [formControl]="form.controls['editor']"></quill-editor>
</form>

<!--<h3>Formula & image resize editor & Keybinding on 'b'</h3>-->
<!--<quill-editor #editor [modules]="modules" (onEditorCreated)="addBindingCreated($event)"></quill-editor>-->

<h3>Bubble editor
  <button type="button" (click)="placeholder=placeholder + '!'">Change placeholder</button>
</h3>
<quill-editor theme="bubble" [placeholder]="placeholder"></quill-editor>

<h3>Editor without toolbar + required and ngModule</h3>
<button (click)="toggleReadOnly()">Toggle ReadOnly</button>
{{isReadOnly}}
{{title}}
<quill-editor [(ngModel)]="title" [maxLength]="5" [minLength]="3" [required]="true" [readOnly]="isReadOnly"
              [modules]="{toolbar: false}" (onContentChanged)="logChange($event);"
              (onSelectionChanged)="logSelection($event);"></quill-editor>
<!-- <h3>Custom Toolbar with toolbar title-attributes + Word counter</h3> -->
<!-- <quill-editor [modules]="{ counter: { container: '#counter', unit: 'word' } }"> -->
  <!--<div quill-editor-toolbar>-->
<!--<span class="ql-formats">-->
<!--<select class="ql-font">-->
<!--<option value="aref">Aref Ruqaa</option>-->
<!--<option value="mirza">Mirza</option>-->
<!--<option selected>Roboto</option>-->
<!--</select>-->
<!--<select class="ql-align" [title]="'Aligment'">-->
<!--<option selected></option>-->
<!--<option value="center"></option>-->
<!--<option value="right"></option>-->
<!--<option value="justify"></option>-->
<!--</select>-->
<!--<select class="ql-align" [title]="'Aligment2'">-->
<!--<option selected></option>-->
<!--<option value="center"></option>-->
<!--<option value="right"></option>-->
<!--<option value="justify"></option>-->
<!--</select>-->
<!--</span>-->
    <!--<span class="ql-formats">-->
<!--<div id="counter"></div>-->
<!--</span>-->
  <!--</div>-->
<!--</quill-editor>-->
```
counter.ts
```
import * as Quill from 'quill';

export interface Config {
  container: string;
  unit: 'word'|'char';
}

export interface QuillInstance {
  on: Function;
  getText: Function;
}

export default class Counter {
  quill: QuillInstance;
  options: Config;

  constructor(quill, options) {
    this.quill = quill;
    this.options = options;

    const container = document.querySelector(this.options.container);

    this.quill.on('text-change', () => {
      const length = this.calculate();

      container.innerHTML = length + ' ' + this.options.unit + 's';
    });
  }

  calculate() {
    const text = this.quill.getText().trim();

    if (this.options.unit === 'word') {
      return !text ? 0 : text.split(/\s+/).length;
    }
    return text.length;
  }
}
```

counter.d.ts although I just added this as a try everything situation, still dont have great understanding of d.ts
```
import * as Quill from 'quill';

export interface Config {
  container: string;
  unit: 'word'|'char';
}

export interface QuillInstance {
  on: Function;
  getText: Function;
}

export default class Counter {
  quill: QuillInstance;
  options: Config;

  constructor(quill, options) {
    this.quill = quill;
    this.options = options;

    const container = document.querySelector(this.options.container);

    this.quill.on('text-change', () => {
      const length = this.calculate();

      container.innerHTML = length + ' ' + this.options.unit + 's';
    });
  }

  calculate() {
    const text = this.quill.getText().trim();

    if (this.options.unit === 'word') {
      return !text ? 0 : text.split(/\s+/).length;
    }
    return text.length;
  }
}
```

**Comment by KillerCodeMonkey** [comment]
Image resize needs quill globally on the window object. Check the webpack
conf in the ngx-example-repo ;)

Am 24.02.2018 10:38 schrieb "giorgio19" <notifications@github.com>:

**Comment by giorgio19** [comment]
Oh my project doesn't have a web pack config, I shall try that.

**Comment by giorgio19** [comment]
:( okay adding the web pack config still didn't solve it. I thought maybe it needs to be specifically in app.component but trying that also didn't work. I am trying to rebuild and reorganize and see if maybe that works. So far just copy and pasting webpack from ngx-quill-example did not solve the problem. Will keep trying, maybe I added it wrong. I need to look into web pack.

**Comment by giorgio19** [comment]
OH haha definitely just copy and pasting is not correct, just looking at web pack documentation and realizing I need to do a couple things most likely. Will try that and then get back.

**Comment by giorgio19** [comment]
Okay so was able to install webpack using documentation on webpack's site 
I did in terminal: 
```npm install --save-dev typescript ts-node @types/node @types/webpack```
which generated the same webpack.config.js you had. But still nothing is generated when I run the program. I tried installing all of the dependencies and devDependencies in your package.json file and that just made a huge mess. 

Also tried `npm install --save-dev webpack` that also didn't work nor generated the appropiate webpack.config.js.

One thing I noticed that our package.json scripts differ. We use `ng serve` to start the program, could that be the key problem? I feel like I am missing something real obvious because I am just working in unfamiliar territory haha, but I also feel like its close.

CLOSEST, I got was trying to match your webpack version `npm install --save-dev webpack@2.3.2 webpack-dev-server@2.9.4 typescript@2.4.2 @types/node @types/webpack`. I dont get any errors but nothing is outputted. :(

**Comment by KillerCodeMonkey** [comment]
There is no need to restructure your project, just check if you van define
globales with angular cli.

The Main Problem is, that image resize is not correctly implemented to use
it with modular JS / typescript.

Greetz.



Am 26.02.2018 04:10 schrieb "giorgio19" <notifications@github.com>:

Okay so was able to install webpack using documentation on webpack's site
I did in terminal:
npm install --save-dev typescript ts-node @types/node @types/webpack
which generated the same webpack.config.js you had. But still nothing is
generated when I run the program. I tried installing all of the
dependencies and devDependencies in your package.json file and that just
made a huge mess.

Also tried npm install --save-dev webpack that also didn't work nor
generated the appropiate webpack.config.js.

One thing I noticed that our package.json scripts differ. We use ng serve
to start the program, could that be the key problem? I feel like I am
missing something real obvious because I am just working in unfamiliar
territory haha, but I also feel like its close.

CLOSEST, I got was trying to match your webpack version npm install
--save-dev webpack@2.3.2 webpack-dev-server@2.9.4 typescript@2.4.2
@types/node @types/webpack. I dont get any errors but nothing is outputted.
:(

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/KillerCodeMonkey/ngx-quill-example/issues/6#issuecomment-368378025>,
or mute the thread

**Comment by giorgio19** [comment]
Ahhh that makes sense will look into that and get back to you. :)

**Comment by giorgio19** [comment]
Meh making a mess of things, now says cannot read property 'register' of undefined, so now its having a problem with Quill.register line. I'm focusing on just trying to do counter as I really have no clue what to do with image resize-module. Gonna try reverting back and trying again haha.

**Comment by giorgio19** [comment]
okay yes I think it has something to do with importing Quill into the component so that Quill.register works. Now that I am focusing on just counter, I get the error Uncaught TypeError: Cannot read property 'register' of undefined.

**Comment by giorgio19** [comment]
I DID ITTTTTTT

**Comment by giorgio19** [comment]
needed in my component.ts file 
```
import * as QuillNamespace from 'quill';
let Quill: any = QuillNamespace;

import Counter from './counter';
Quill.register('modules/counter', Counter);
```

**Comment by KillerCodeMonkey** [comment]
Congrats :) Nice you got it!

Close the issue if it is solved ;)

**Comment by giorgio19** [comment]
Thank you!!

**Comment by giorgio19** [comment]
also you can do this without acutally having to npm install webpack like i said before, as long as you add that line of code it all works with my package.json just being this
```
{
  "name": "testing",
  "version": "0.0.0",
  "license": "MIT",
  "scripts": {
    "ng": "ng",
    "start": "ng serve",
    "build": "ng build --prod",
    "test": "ng test",
    "lint": "ng lint",
    "e2e": "ng e2e"
  },
  "private": true,
  "dependencies": {
    "@angular-devkit/core": "0.0.29",
    "@angular/animations": "^5.1.0",
    "@angular/common": "^5.2.1",
    "@angular/compiler": "^5.2.1",
    "@angular/core": "^5.2.1",
    "@angular/forms": "^5.1.0",
    "@angular/http": "^5.1.0",
    "@angular/platform-browser": "^5.2.1",
    "@angular/platform-browser-dynamic": "^5.2.1",
    "@angular/router": "^5.1.0",
    "@ng-bootstrap/ng-bootstrap": "1.0.0-beta.9",
    "angularfire": "^2.3.0",
    "angularfire2": "^5.0.0-rc.6",
    "bootstrap": "4.0.0-beta.2",
    "core-js": "^2.4.1",
    "firebase": "^4.9.0",
    "jquery": "^3.3.1",
    "mathjax": "^2.7.2",
    "ngx-quill": "^2.1.1",
    "popper.js": "^1.12.9",
    "quill": "^1.3.5",
    "rxjs": "^5.5.6",
    "zone.js": "^0.8.19"
  },
  "devDependencies": {
    "@angular-devkit/core": "0.0.28",
    "@angular/cli": "1.6.4",
    "@angular/compiler-cli": "^5.1.0",
    "@angular/language-service": "^5.1.0",
    "@types/jasmine": "^2.8.5",
    "@types/jasminewd2": "~2.0.2",
    "@types/node": "~6.0.60",
    "codelyzer": "^4.0.1",
    "jasmine-core": "~2.8.0",
    "jasmine-spec-reporter": "~4.2.1",
    "karma": "~2.0.0",
    "karma-chrome-launcher": "~2.2.0",
    "karma-cli": "~1.0.1",
    "karma-coverage-istanbul-reporter": "^1.4.0",
    "karma-jasmine": "~1.1.0",
    "karma-jasmine-html-reporter": "^0.2.2",
    "protractor": "~5.1.2",
    "ts-node": "~3.2.0",
    "tslint": "~5.9.1",
    "typescript": "~2.5.3"
  }
}
```

**Comment by giorgio19** [comment]
also worked with only counter.ts instead of needing that and counter.d.ts

**Comment by Maddy2894** [comment]
Hey @giorgio19 I am not using webpack and followed your code, I am getting this error, 
<img width="512" alt="screen shot 2019-02-03 at 11 43 46 pm" src="https://user-images.githubusercontent.com/42700140/52176900-a6d25980-280d-11e9-8645-abc61e8a724f.png"> Any leads?

**Comment by KillerCodeMonkey** [comment]
Are you using angular cli?

Because you mentioned Not using webpack, but in your Screenshot webpack
Source is shown :)

Am So., 3. Feb. 2019, 13:44 hat Madhavan Sundararaj <
notifications@github.com> geschrieben:

**Comment by Maddy2894** [comment]
No, i am using angular.json

**Comment by Maddy2894** [comment]
Anyways, I want the uploaded image to be of width 100% by default in quill js editor, any method of doing it. If it's possible, I do not need image resize module. 

Cheers.

**Comment by KillerCodeMonkey** [comment]
My example Repo Shows how to use Image resize Module. And yes you can Style
IT with CSS. Just checkout, what quilljs adds when adding an Image and
Style IT in your global Styling.

Am So., 3. Feb. 2019, 14:16 hat Madhavan Sundararaj <
notifications@github.com> geschrieben:

**Comment by Maddy2894** [comment]
I tried doing it, but no changes are shown, please tell me a way to set image width to 100%. That's enough.

**Comment by KillerCodeMonkey** [comment]
i will not solve your quilljs styling issue for you.

Hint: you have to set it in your global styling not in your component, because angular adds ViewEncapsulation by default :)

**Comment by Maddy2894** [comment]
Haha. Thank you for your hint. I worked it out. Thanks a lot.

Cheers.

**Comment by Maddy2894** [comment]
Just a doubt, I want to keep your default toolbar options and add one more like counter, is there a way to concat or I should list down all the options?

**Comment by KillerCodeMonkey** [comment]
Released a new version 4.6.1 some seconds ago, were QuillModules exports the defaultModules.

So you can use it to extend it on your own :)

**Comment by Maddy2894** [comment]
Cool!

**Comment by KillerCodeMonkey** [comment]
maybe in a next version

**Comment by vinodchauhan99** [comment]
Quill Mentions - How can add user image in the mention list?

**Comment by KillerCodeMonkey** [comment]
ask at the quill mention repo :)

**Comment by 4gmz** [comment]
estoy utlizando vue.js y tengo el siguiente error :
 quill.js?9339:1072 Uncaught TypeError: **Cannot read property 'register' of undefined**
    at Function.register (quill.js?9339:1072)
    at eval (quill.js?9980:8)
    at Module../resources/js/plugins/quill.js (app.js:7305)
    at __webpack_require__ (app.js:64)
    at eval (index.js?af27:1)
    at Module../resources/js/plugins/index.js (app.js:7269)
    at __webpack_require__ (app.js:64)
    at eval (app.js:9)
    at Module../resources/js/app.js (app.js:6411)
    at __webpack_require__ (app.js:64)

de mi archivo .js 
import Vue from 'vue'
import VueQuill from 'vue-quill'
import Quill from 'quill'
import { ImageResize } from 'quill-image-resize-module'

Vue.use(VueQuill)
Quill.register('modules/imageResize', ImageResize)

**Comment by KillerCodeMonkey** [comment]
@4gmz 

1. in english please
2. you are using VueJS and not ngx-quill
3. register of undefined means, that Quill is not defined.

so in general i would import vue-quill after the import of quill, and try importing Quill like mentioned above

`import * as Quill from 'quill';`

**Comment by Despicable-bee** [comment]
hi, sorry to reopen this thread, but I figure it's the best place for it.

I'm having the same issue as @giorgio19 where trying to get the counter.ts module integrated as an example keeps raising:

`Uncaught (in promise): TypeError: Cannot read property 'register' of null`

I noticed some peculiar behaviour while stepping through the chrome debugger for the past 4 hours.

I get this warning right before the TypeError:

`quill Overwriting modules/counter with null` on line `quill.js: 2037`

However, when I include the following:

```
import * as QuillNamespace from 'quill';
let Quill: any = QuillNamespace;

import Counter from './counter';
Quill.register('modules/counter', Counter);
```

in BOTH the `editor.components.ts` file, and in the `app.components.ts` file. I get:

```
quill Overwriting modules/counter with class Counter {
    constructor(quill, options) {
        this.quill = quill;
        this.options = options;
        const container = document.querySelector(this.options.container);
        this.quill.on…
```

followed by

`quill Overwriting modules/counter with null` on line `quill.js: 2037`

I have absolutely no idea why this happens, and I'd really appreciate some help :)

I've spent quite a while trying to configure the ckeditor5 to my needs, but its upload adapters are confusing to work with, and documentation isn't very clear on how it works.

Kind regards

Despicable b
